presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/rtree/Flatbush.java

```diff
@@ -0,0 +1,320 @@
+ * object's Hilbert curve index.  Hilbert curves are naturally hierarchical, so
+ * successively grouping children and their parents will give a naturally nested
+ * structure.  This means we only need to sort the items once.
+ * <p>
+ * If sort time is a problem, we actually just need to "partition" into groups
+ * of `degree`, since the order of the children of a single parent doesn't
+ * matter.  This could be done with quicksort, stopping once all items at index
+ * `n * degree` are correctly placed.
+ * <p>
+ * Original implementation in Javascript: https://github.com/mourner/flatbush
+ */
+public class Flatbush<T extends HasExtent>
+{
+    private static final int FLATBUSH_INSTANCE_SIZE = ClassLayout.parseClass(Flatbush.class).instanceSize();
+    private static final int DEFAULT_DEGREE = 16;
+    // Number of coordinates to define an envelope
+    private static final int ENVELOPE_SIZE = 4;
+    // How many children per node
+    private final int degree;
+    // Offsets in tree for each level
+    private final int[] levelOffsets;
+    // Each node will have four doubles: xMin, yMin, xMax, yMax
+    private final DoubleBigArray tree;
+    private final T[] items;
+
+    public Flatbush(T[] items, int degree)
+    {
+        checkArgument(degree > 0, "degree must be positive");
+        this.degree = degree;
+        this.items = requireNonNull(items, "items is null");
+        this.levelOffsets = Flatbush.calculateLevelIndices(items.length, degree);
+        this.tree = this.buildTree();
+    }
+
+    public Flatbush(T[] items)
+    {
+        this(items, DEFAULT_DEGREE);
+    }
+
+    /* Calculate the indices for each level.
+     *
+     * A given level contains a certain number of items.  We give it a capacity
+     * equal to the next multiple of `degree`, so that each parent will have
+     * an equal number (`degree`) of children.  This means the next level will
+     * have `capacity/degree` nodes, which yields a capacity equal to the next
+     * multiple, and so on, until the number of items is 1.
+     *
+     * Since we are storing each node as 4 doubles, we will actually multiply
+     * every index by 4.
+     */
+    private static int[] calculateLevelIndices(int numItems, int degree)
+    {
+        List<Integer> levelIndices = new ArrayList<Integer>();
+        // Leaf nodes start at 0, root is the last element.
+        levelIndices.add(Integer.valueOf(0));
+        int capacity;
+        int level = 0;
+        while (numItems > 1) {
+            capacity = Flatbush.findNextMultiple(numItems, degree);
+            levelIndices.add(levelIndices.get(level) + ENVELOPE_SIZE * capacity);
+            numItems = capacity / degree;
+            level += 1;
+        }
+        return levelIndices.stream().mapToInt(Integer::intValue).toArray();
+    }
+
+    /**
+     * Find the least multiple of `factor` gte to `n`.
+     * <p>
+     * n must be positive.
+     * TODO: Probably put this in utilities somewhere?
```

Comment:
I'd inline this method and change it to something like `int capacity = (int) Math.ceil(numItems / degree) * degree;`

Reply:
While I haven't figured out why, the `Math.ceil` approach is giving a different answer and making the tests fail.  I've inlined the integer computation and put in an explanatory comment.